provisioning/transport/amqp/src/AmqpTransportHandler.cs

```diff
@@ -0,0 +1,145 @@
+
+            TpmNetworkCredential credential = new TpmNetworkCredential(ek, srk);
+
+                var saslProvider = new SaslTransportProvider();
+                saslProvider.Versions.Add(AmqpConstants.DefaultProtocolVersion);
+                settings.TransportProviders.Add(saslProvider);
+
+            SaslTpmHandler tpmHandler = new SaslTpmHandler(ekBuffer, srkBuffer, linkendpoint, securityClient);
+            saslProvider.AddHandler(tpmHandler);
+
+            var amqpProvider = new AmqpTransportProvider();
+            amqpProvider.Versions.Add(AmqpConstants.DefaultProtocolVersion);
+            settings.TransportProviders.Add(amqpProvider);
+            return settings;
+        }
+
+        internal static async Task<AmqpClientConnection> CreateAmqpCloudConnectionAsync(
+            string deviceEndpoint, 
+            X509Certificate2 clientCert, 
+            string linkEndpoint, 
+            bool useWebSocket,
+            ProvisioningSecurityClientSasToken securityClient)
+        {
+            AmqpClientConnection amqpCloudConnection;
+            if (useWebSocket)
+            {
+                // TODO: enable WS
+                amqpCloudConnection = await CreateConnection(
+                    new Uri(WebSocketConstants.Scheme + deviceEndpoint + ":" + WebSocketConstants.Port),
+                    securityClient,
+                    linkEndpoint).ConfigureAwait(false);
+        }
+        else
+            {
+                amqpCloudConnection = await CreateConnection(
+                    new Uri("amqps://" + deviceEndpoint + ":" + AmqpConstants.DefaultSecurePort), 
+                    securityClient, 
+                    linkEndpoint).ConfigureAwait(false);
+            }
+
+            await amqpCloudConnection.OpenAsync(TimeSpan.FromMinutes(1), useWebSocket, clientCert).ConfigureAwait(false);
+
+            var amqpDeviceSession = amqpCloudConnection.CreateSession();
+            await amqpDeviceSession.OpenAsync(TimeSpan.FromMinutes(1)).ConfigureAwait(false);
+
+            var amqpReceivingLink = amqpDeviceSession.CreateReceivingLink(linkEndpoint);
+
+            amqpReceivingLink.AddClientVersion(clientVersion);
+            amqpReceivingLink.AddApiVersion(apiVersion);
+
+            await amqpReceivingLink.OpenAsync(TimeSpan.FromMinutes(1)).ConfigureAwait(false);
+
+            var amqpSendingLink = amqpDeviceSession.CreateSendingLink(linkEndpoint);
+
+            amqpSendingLink.AddClientVersion(clientVersion);
+            amqpSendingLink.AddApiVersion(apiVersion);
+
+            await amqpSendingLink.OpenAsync(TimeSpan.FromMinutes(1)).ConfigureAwait(false);
+
+            return amqpCloudConnection;
+        }
+
+        public static async Task<RegistrationOperationStatus> RegisterDeviceAsync(AmqpClientConnection client, string correlationId)
+        {
+            var amqpMessage = AmqpMessage.Create(new MemoryStream(Encoding.ASCII.GetBytes(DeviceOperations.Register)), false);
```

Comment:
Seems like overhead that we need to start the message with the "iotdps-register" string then add the same string within the properties.